### PR TITLE
Improve Rule Generation

### PR DESCRIPTION
**Pull request by mjohnson541**
This has four main changes
1) Enables tree generation for charge transfer families
2) Updates tree generation, if E0 < 0 or n = 1 it simply uses the average instead of the BM rule
3) Improves BM fitting using dH298
4) Adds a SolventData attribute to reactions that can hold transition state solute information

### CODE REVIEW

File: rmgpy/reaction.py [hunk 1]
```diff
@@ -880,9 +877,60 @@ def reverse_sticking_coeff_rate(self, k_forward, reverse_units, surface_site_den
         kr.fit_to_data(Tlist, klist, reverse_units, kf.T0.value_si)
         return kr
 
+<<<<<<< HEAD
```
Comment (hwpang) [inline rmgpy/reaction.py:880]: Can you resolve this? It seems some version conflicts are commited

File: rmgpy/reaction.py [hunk 2]
```diff
@@ -880,9 +877,60 @@ def reverse_sticking_coeff_rate(self, k_forward, reverse_units, surface_site_den
...
+
+        V0 = self.kinetics.V0.value_si
+        klist = np.zeros_like(Tlist)
+        for i in range(len(Tlist)):
+            klist[i] = kf.get_rate_coefficient(Tlist[i],V0) / self.get_equilibrium_constant(Tlist[i],V0)
+        kr = ArrheniusChargeTransfer(alpha=kf.alpha.value, electrons=-1*self.electrons, V0=(V0,'V'))
+        kr.fit_to_data(Tlist, klist, reverse_units, kf.T0.value_si)
+        return kr
+
+>>>>>>> 3afc9cc91 (handle ChargeTransfer kinetics with in reaction.py)
```
Comment (hwpang) [inline rmgpy/reaction.py:930]: Could you remove this

File: rmgpy/reaction.py [hunk 3]
```diff
@@ -905,22 +953,34 @@ def generate_reverse_rate_coefficient(self, network_kinetics=False, Tmin=None, T
...
-            logging.warning(f"Species do not have an rmgpy.molecule.Molecule "  
+            logging.warning(f"Species do not have an rmgpy.molecule.Molecule "
                             "Cannot determine phases of species. We will assume gas"
                             )
         n_surf = len(surf_prods)
         n_gas = len(self.products) - len(surf_prods)
         kunits = get_rate_coefficient_units_from_reaction_order(n_gas, n_surf)
 
         kf = self.kinetics
+<<<<<<< HEAD
```
Comment (hwpang) [inline rmgpy/reaction.py:972]: Could you remove this

File: rmgpy/reaction.py [hunk 4]
```diff
@@ -905,22 +953,34 @@ def generate_reverse_rate_coefficient(self, network_kinetics=False, Tmin=None, T
...
+=======
+
+        if isinstance(kf, SurfaceChargeTransfer):
+            return self.reverse_surface_charge_transfer_rate(kf, kunits, Tmin, Tmax)
+
+        elif isinstance(kf, ArrheniusChargeTransfer):
+            return self.reverse_arrhenius_charge_transfer_rate(kf, kunits, Tmin, Tmax)
+
+        elif isinstance(kf, KineticsData):
+>>>>>>> 3afc9cc91 (handle ChargeTransfer kinetics with in reaction.py)
```
Comment (hwpang) [inline rmgpy/reaction.py:983]: Could you remove this and resolve the conflicts?

File: rmgpy/reaction.py [hunk 5]
```diff
@@ -1118,14 +1178,22 @@ def is_balanced(self):
...
                 return False
 
+<<<<<<< HEAD
+=======
+        if self.electrons < 0:
+            reactants_net_charge += self.electrons
+        elif self.electrons > 0:
+            products_net_charge -= self.electrons
+
+>>>>>>> 3afc9cc91 (handle ChargeTransfer kinetics with in reaction.py)
```
Comment (hwpang) [inline rmgpy/reaction.py:1188]: Could you remove this and resolve the conflicts?

File: rmgpy/data/kinetics/family.py [hunk 6]
```diff
@@ -3704,47 +3704,43 @@ def cross_validate(self, folds=5, template_rxn_map=None, test_rxn_inds=None, T=1
...
-                if not ascend:
-                    L = list(set(template_rxn_map[entry.label]) - set(rxns_test))
 
-                    if L != []:
-                        if isinstance(L[0].kinetics,Arrhenius):
-                            kinetics = ArrheniusBM().fit_to_reactions(L, recipe=self.forward_recipe.actions)
-                            kinetics = kinetics.to_arrhenius(rxn.get_enthalpy_of_reaction(T))
+                L = list(set(template_rxn_map[entry.label]) - set(rxns_test))
+
+                if L != []:
```
Comment (hwpang) [inline rmgpy/data/kinetics/family.py:3726]: I think a more Pythonic way to check empty list is by `if not L`?

File: rmgpy/data/kinetics/family.py [hunk 7]
```diff
@@ -4602,25 +4602,61 @@ def _make_rule(rr):
...
+                ws = 1.0 / varis
+                V1 = ws.sum()
+                V2 = (ws ** 2).sum()
+                mu = np.dot(ws, dlnks) / V1
+                s = np.sqrt(np.dot(ws, (dlnks - mu) ** 2) / (V1 - V2 / V1))
+                kin.uncertainty = RateUncertainty(mu=mu, var=s ** 2, N=n, Tref=Tref, data_mean=data_mean, correlation=label)
+                return kin
+        else:
+            if n == 1:
+                kin.uncertainty = RateUncertainty(mu=0.0, var=(np.log(fmax) / 2.0) ** 2, N=1, Tref=Tref, data_mean=data_mean, correlation=label)
```
Comment (rwest) [inline rmgpy/data/kinetics/family.py:4654]: I think this is unreachable? 
```
if n==1: # on line 4611
else: # line 4631
    if n==1: # line 432
```

File: rmgpy/data/kinetics/family.py [hunk 8]
```diff
@@ -3704,47 +3704,43 @@ def cross_validate(self, folds=5, template_rxn_map=None, test_rxn_inds=None, T=1
                     if entry.parent:
                         entry = entry.parent
 
+                boo = True
```
Comment (rwest) [inline rmgpy/data/kinetics/family.py:3707]: Why `boo`? A more informative name would help. ANd I don't see where `boo` is ever set to anything but True? So what's this infinite loop about? 
Does it mean `while entry.parent:` ?

File: rmgpy/kinetics/arrhenius.pyx [hunk 9]
```diff
@@ -611,27 +611,24 @@ cdef class ArrheniusBM(KineticsModel):
...
             Ea = rxn.kinetics.Ea.value_si
-            
+
             def kfcn(E0):
                 Vp = 2 * w0 * (2 * w0 + 2 * E0) / (2 * w0 - 2 * E0)
                 out = Ea - (w0 + dHrxn / 2.0) * (Vp - 2 * w0 + dHrxn) * (Vp - 2 * w0 + dHrxn) / (Vp * Vp - (2 * w0) * (2 * w0) + dHrxn * dHrxn)
                 return out
 
-            if abs(dHrxn) > 4 * w0 / 10.0:
-                E0 = w0 / 10.0
```
Comment (rwest) [inline rmgpy/kinetics/arrhenius.pyx:614]: can you explain why removing this? (in the commit message, ideally)

**Review by rwest — COMMENTED** [review]
Need to figure out how this relates to https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2264
I think both are trying to solve some similar problems.

Can you explain why NOT to fit a Blowers Masel when there's only one training reaction? (if that's the new proposal) ?

**Comment by mjohnson541** [comment]
I'm not sure how up to date this PR is relative to the electrochem branch. It's been a while, I think as defined in the paper the dHrxn is at 298 K which implies that the Ea is constant with respect to temperature so you only have one datapoint to determine E0. You can determine E0 this way, but I think someone was having problems with those fits? At some point I think this was a matter of significant urgency to get merged and then we (or at least I) forgot about it?

**Comment by rwest** [comment]
I think this would address https://github.com/ReactionMechanismGenerator/RMG-Py/issues/1748

**Comment by rwest** [comment]
I think this is now part of https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2316 which has been more recently rebased and checked. Hopefully that will be merged soon.

**Pull request closed without merging**